plugins/unreal/XJGame/Plugins/XJMusicPlugin/Source/XJMusicPlugin/Private/XjMusicInstanceSubsystem.cpp

```diff
@@ -0,0 +1,188 @@
+	{
+		return;
+	}
+
+	TArray<FString> FoundAudioFilesPaths;
+
+	PlatformFile.FindFilesRecursively(FoundAudioFilesPaths, *WorkPath, *AudioExtension);
+
+	for (const FString& Path : FoundAudioFilesPaths)
+	{
+		FString Name = FPaths::GetCleanFilename(Path);
+		AudioPathsByNameLookup.Add(Name, Path);
+	}
+}
+
+void UXjMusicInstanceSubsystem::PlayAudioByName(const FString& Name, const float OverrideStartTime)
+{
+	if (SoundsMap.Contains(Name))
+	{
+		return;
+	}
+
+	USoundWave* SoundWave = GetSoundWaveByName(Name);
+	if (!SoundWave)
+	{
+		return;
+	}
+
+	if (FAudioDeviceHandle Device = GetWorld()->GetAudioDevice())
+	{
+		TSharedPtr<FActiveSound> ActiveSound = MakeShared<FActiveSound>();
+		ActiveSound->SetSound(SoundWave);
+		ActiveSound->SetWorld(GetWorld());
+
+		ActiveSound->SetVolume(1.0f);
+
+		ActiveSound->RequestedStartTime = OverrideStartTime;
+
+		ActiveSound->bIsUISound = true;
+		ActiveSound->bAllowSpatialization = false;
+
+		ActiveSound->Priority = SoundWave->GetPriority();
+		ActiveSound->SubtitlePriority = SoundWave->GetSubtitlePriority();
+
+		SoundsMap.Add(Name, ActiveSound);
+
+		Device->AddNewActiveSound(*ActiveSound.Get());
+	}
+}
+
+void UXjMusicInstanceSubsystem::StopAudioByName(const FString& Name)
+{
+	if (!SoundsMap.Contains(Name))
+	{
+		return;
+	}
+
+	FActiveSound* Sound = SoundsMap[Name].Get();
+
+	if (!Sound)
+	{
+		return;
+	}
+
+	FAudioThread::RunCommandOnAudioThread([this, Sound, Name]()
+		{
+			TArray<FActiveSound*> ActiveSounds = WorldAudioDeviceHandle->GetActiveSounds();
+
+			for (int32 SoundIndex = ActiveSounds.Num() - 1; SoundIndex >= 0; --SoundIndex)
+			{
+				FActiveSound* Inst = ActiveSounds[SoundIndex];
+				if (Inst && (Inst->GetSound()->GetName() == Sound->GetSound()->GetName()))
+				{
+					WorldAudioDeviceHandle->AddSoundToStop(Inst);
+					SoundsMap.Remove(Name);
+
+					break;
+				}
+			}
+		});
+}
+
+USoundWave* UXjMusicInstanceSubsystem::GetSoundWaveByName(const FString& AudioName)
+{
+	if (!AudioPathsByNameLookup.Contains(AudioName))
+	{
+		return nullptr;
+	}
+
+	FString FilePath = AudioPathsByNameLookup[AudioName];
+
+	USoundWave* SoundWave = NewObject<USoundWave>(USoundWave::StaticClass());
+	if (!SoundWave)
+	{
+		UE_LOG(LogTemp, Error, TEXT("Failed to create USoundWave object"));
+		return nullptr;
+	}
+
+	TArray<uint8> RawFile;
+	if (!FFileHelper::LoadFileToArray(RawFile, *FilePath))
+	{
+		return nullptr;
+	}
+
+	SoundWave->SoundGroup = ESoundGroup::SOUNDGROUP_Default;
+	SoundWave->NumChannels = 2; 
+
+	SoundWave->RawData.Lock(LOCK_READ_WRITE);
+
+	void* LockedData = SoundWave->RawData.Realloc(RawFile.Num());
+	if (LockedData)
+	{
+		FMemory::Memcpy(LockedData, RawFile.GetData(), RawFile.Num());
+	}
+
+	SoundWave->RawData.Unlock();
+
+	SoundWave->RawData.SetBulkDataFlags(BULKDATA_ForceInlinePayload);
+	SoundWave->InvalidateCompressedData();
+	//SoundWave->bNeedsThumbnailGeneration = true;
+
+	return SoundWave;
+}
+
+void UXjMusicInstanceSubsystem::TestPlayAllSounds()
```

Comment:
Since this method serves the test harness XjTestActor my preference is for all the non-production methods to be moved to XjTestActor. If we create a method on XjMusicInstanceSubsystem it should be agnostic to whether it serves a production or test scenario.

Reply:
I agree with you. I will just remove TestActor since I doens't need it more